### PR TITLE
Return Skip in PodTopologySpread#PreScore under specific conditions

### PR DESCRIPTION
**Pull request by utam0k**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Return Skip in PodTopologySpread#PreScore under specific conditions.

#### Which issue(s) this PR fixes:

This change is a part of the work for issue #115745

#### Does this PR introduce a user-facing change?

```release-note
The scheduler skips the PodTopologySpread Score plugin when nothing to do with the Pod.
It will affect some metrics values related to the PodTopologySpread Score plugin.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Review by sanposhiho — COMMENTED** [review]
/ok-to-test

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 1]
```diff
@@ -42,6 +42,101 @@ import (
...
+			pl, err := New(&tt.config, f, feature.Features{})
+			if err != nil {
+				t.Fatalf("Failed creating plugin: %v", err)
+			}
+			informerFactory.Start(ctx.Done())
+			informerFactory.WaitForCacheSync(ctx.Done())
+			p := pl.(*PodTopologySpread)
+			cs := framework.NewCycleState()
+			if s := p.PreScore(context.Background(), cs, tt.pod, tt.nodes); !s.IsSkip() {
+				t.Fatal(s.AsError())
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:134]: Please make it a richer error message like "expect skip but got ...

**Comment by SataQiu** [comment]
/triage accepted

Hi @utam0k The pull-kubernetes-integration test need to be fixed.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go [hunk 2]
```diff
@@ -115,14 +115,18 @@ func (pl *PodTopologySpread) PreScore(
 	pod *v1.Pod,
 	filteredNodes []*v1.Node,
 ) *framework.Status {
+	if len(filteredNodes) == 0 {
```
[the remark this answers is not in the document]
Comment (SataQiu) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring.go:118]: That is true from https://github.com/kubernetes/kubernetes/blob/232cdf97164f6e3a77ee954bbdd29866232f04dc/pkg/scheduler/schedule_one.go#L360-L366

But it looks like we already did the check in some `PreScore` plugin, such as https://github.com/kubernetes/kubernetes/blob/232cdf97164f6e3a77ee954bbdd29866232f04dc/pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go#L186-L188

Do we need to clean up all of this code? maybe I can help.

**Comment by sanposhiho** [comment]
@utam0k Please rebase this branch. I believe that'll fix the integration test failing.

**Comment by sanposhiho** [comment]
/assign
so that it won't slip my mind.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 3]
```diff
@@ -42,6 +42,101 @@ import (
...
+			objs: []runtime.Object{
+				&appsv1.ReplicaSet{ObjectMeta: metav1.ObjectMeta{Namespace: "default", Name: "rs1"}, Spec: appsv1.ReplicaSetSpec{Selector: fooSelector}},
+			},
+		},
+		{
+			name: "the pod doesn't have soft topology spread Constraints",
+			pod:  st.MakePod().Name("p").Namespace("default").Obj(),
+			config: config.PodTopologySpreadArgs{
+				DefaultingType: config.ListDefaulting,
+			},
```
[the remark this answers is not in the document]
Comment (utam0k) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:83]: No, it seems that we need to set `System` or `List`
```
    --- FAIL: TestPreScoreSkip/the_pod_doesn't_have_soft_topology_spread_Constraints (0.00s)
        scoring_test.go:101: Failed creating plugin: defaultingType: Unsupported value: "": supported values: "System", "List"
```

**Comment by utam0k** [comment]
@sanposhiho Thanks for your review 🙇 I have updated and rebased it.

**Review by sanposhiho — COMMENTED** [review]
small nit, will give /lgtm once it's done

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 4]
```diff
@@ -42,6 +42,74 @@ import (
...
+			}
+			pl, err := New(&tt.config, f, feature.Features{})
+			if err != nil {
+				t.Fatalf("Failed creating plugin: %v", err)
+			}
+			informerFactory.Start(ctx.Done())
+			informerFactory.WaitForCacheSync(ctx.Done())
+			p := pl.(*PodTopologySpread)
+			cs := framework.NewCycleState()
+			if s := p.PreScore(context.Background(), cs, tt.pod, tt.nodes); !s.IsSkip() {
```
[the remark this answers is not in the document]
Comment (utam0k) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:106]: Thanks for your review. Done.

[the remark this answers is not in the document]
Comment (utam0k) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:106]: https://github.com/kubernetes/kubernetes/compare/c0a11415954e74fd01ec75cdeaf47364451993df..ef2651016482195d5cc5439b15bd28b9f6c396f1

**Review by sanposhiho — COMMENTED** [review]
Thanks
/lgtm

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Pull request merged**